File: problem2.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// the alogrithm iterates backward through nums array and calculate max amount that can be robbed
// calculates the maximum amount that can be robbed without adjacent houses using
// an iterative bottom-up approach
class problem2 {

    public int rob(int[] nums) {

        int N = nums.length;

        if (N == 0) {
            return 0;
        }

        int[] maxRobbedAmount = new int[nums.length + 1];

        maxRobbedAmount[N] = 0;
        maxRobbedAmount[N - 1] = nums[N - 1];

        for (int i = N - 2; i >= 0; --i) {

            maxRobbedAmount[i] = Math.max(maxRobbedAmount[i + 1], maxRobbedAmount[i + 2] + nums[i]);
        }

        return maxRobbedAmount[0];
    }
}

File: problem2.go

// Time Complexity : O(n)
// Space Complexity : O(1) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


//Approach 1 Time limit exceed

func rob(nums []int) int {
    if len(nums) == 0{
        return 0
    }   
    
    return helper(nums,0,0)
}


func helper(nums []int, index int, amount int) int{
    
    
    if index >= len(nums){
        return amount
    }
    
    case0:= helper(nums,index+1,amount)
    
    case1:= helper(nums,index+2,amount+nums[index])
    
    return max(case0,case1)
}


func max(a,b int) int{
    if a > b{
        return a
    }
    
    return b
}


//Approach 2
func rob(nums []int) int {
    if len(nums) == 0{
        return 0
    } 
 
    n := len(nums)
    
    skip:=0
    take:=nums[0]
    
    for i:=1;i<n;i++{
        temp := skip
        
        skip = max(skip,take)
        take = temp+nums[i]
    }
    
    return max(skip,take)
    
}
